File: practice/sort/counting_sort/counting_sort.py
MAX_NUM = 1000

A = list(map(int, input().split(" ")))
N = len(A)

count = [0]*(MAX_NUM+1)
countSum = [0]*(MAX_NUM+1)

for i in range(0, N):
    count[A[i]] += 1

countSum[0] = count[0]
for i in range(1, MAX_NUM+1):
    countSum[i] = countSum[i-1]+count[i]

B = [0]*(N+1)

for i in range(N-1, -1, -1):
    B[countSum[A[i]]] = A[i]
    countSum[A[i]] -= 1

result = ""
for i in range(1, N+1):
    result += str(B[i]) + ""
print(result)

File: practice/sort/counting_sort/counting_sort.ts
let countingSort = function (a: number[], k: number) {
  let count = [], result: any = [];
  for (let i = 0; i <= k; i++) {
    count[i] = 0;
  }
  console.log(count, result, a.length);
  for (let j = 0; j < a.length; j++) {
    count[a[j]] += 1;
  }
  console.log(count, result, k);
  for (let i = 0; i < k; i++) {
    count[i + 1] += count[i];
  }
  console.log(count, result);
  for (let j = 0; j < a.length; j++) {
    console.log(a[j], count[a[j]] - 1);
    result[count[a[j]] - 1] = a[j];
    count[a[j]] -= 1;
  }
  console.log(count, result);
  return result;
};

countingSort([3, 4, 0, 1, 2, 4, 2, 4], 4);
